agent/lib/src/internal_api/service.rs

```diff
@@ -0,0 +1,306 @@
+    pub fn run(&mut self) -> Result<ResponseMessage> {
+        info!("Running all steps");
+        let mut steps: Vec<ResponseStep> = Vec::new();
+
+        // For each commands in the request, run it
+        let steps_to_process = self.request_message.data.steps.clone();
+
+        for step in steps_to_process {
+            let command = step.command.as_str();
+            let code_return = self.run_one(command)?;
+
+            // Hide Stdout if enable_output is false
+            let stdout = if step.enable_output {
+                code_return.stdout
+            } else {
+                "".to_string()
+            };
+            let response_step = ResponseStep::new(
+                command.to_string(),
+                code_return.exit_code,
+                stdout,
+                code_return.stderr,
+                step.enable_output,
+            );
+
+            steps.push(response_step);
+        }
+
+        let data: ResponseData = ResponseData::new(self.request_message.data.id.clone(), steps);
+        let response_message = ResponseMessage::new(data);
+
+        Ok(response_message)
+    }
+
+    /// Run a command
+    ///
+    /// # Arguments
+    ///
+    /// * `command` - The command to run
+    ///
+    /// # Returns
+    ///
+    /// * `Result<CodeReturn>` - The code return or an error
+    pub fn run_one(&mut self, command: &str) -> Result<CodeReturn> {
+        info!("Running command : {}", command);
+
+        let child_process = Command::new("/bin/sh")
+            .args(["-c", command])
+            .current_dir(WORKSPACE_PATH)
+            .output()
+            .map_err(|e| anyhow!("Failed to spawn command : {}", e))?;
+
+        let exit_code = child_process
+            .status
+            .code()
+            .ok_or_else(|| anyhow!("Failed to retrieve exit_code"))?;
+        let stdout = String::from_utf8(child_process.stdout)
+            .map_err(|e| anyhow!("Failed to retrieve stdout stream : {}", e))?;
+        let stderr = String::from_utf8(child_process.stderr)
+            .map_err(|e| anyhow!("Failed to retrieve stderr stream : {}", e))?;
+
+        let code_return = CodeReturn::new(stdout, stderr, exit_code);
+
+        info!("Code execution finished: {:?}", code_return);
+        Ok(code_return)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::external_api::model::{FileModel, RequestData, RequestStep};
+    use std::fs::File;
+    use std::io::Read;
+
+    /// Generate a random integer
+    ///
+    /// # Arguments
+    ///
+    /// * `max` - The maximum value
+    ///
+    /// # Returns
+    ///
+    /// * `usize` - The random integer
+    fn random_usize(max: usize) -> usize {
+        let mut f = File::open("/dev/urandom").unwrap();
+        let mut buf = [0u8; 1];
+        f.read_exact(&mut buf).unwrap();
+        let value = buf[0] as usize;
+
+        if value < max {
+            max
+        } else {
+            value % max
+        }
+    }
+
```

Comment:
https://docs.rs/rand/latest/rand/

Reply:
I didnt use it as I wasn't aware of the ``[dev-dependencies]`` section. Replaced :+1: